### PR TITLE
Clarify ENV non-recursivenes w.r.t. env var substitution + fix escaping

### LINKED ISSUE #10391: ENV declarations don't inherit values if declared on same line
Attempting to declare multiple variables that reference previously declared variables on one line doesn't seem to work, for example this declaration will create variables that contain strings instead of expanded variable definitions:

    ENV JAVA_VERSION=jdk1.7.0_75 \
        JAVA_HOME=/usr/$JAVA_VERSION \ 
        JAVA_BIN=$JAVA_HOME/bin \
        JRE_HOME=$JAVA_HOME/jre

After building a container with the above ENV declarations the environment will contain:

    JAVA_VERSION=jdk1.7.0_75
    JAVA_HOME='/usr/$JAVA_VERSION'
    JAVA_BIN='$JAVA_HOME/bin'
    JRE_HOME='$JAVA_HOME/jre'

Where I would expect those variables to have been expanded.

On the other hand creating variables using multiple ENV statements will work and variables will contain the expanded variables:

    ENV JAVA_VERSION jdk1.7.0_75
    ENV JAVA_HOME /usr/$JAVA_VERSION
    ENV JAVA_BIN $JAVA_HOME/bin
    ENV JRE_HOME $JAVA_HOME/jre

After building a container with the above ENV declarations the environment will contain:

    JAVA_VERSION=jdk1.7.0_75
    JAVA_HOME=/usr/jdk1.7.0_75
    JAVA_BIN=/usr/jdk1.7.0_75/bin
    JRE_HOME=/usr/jdk1.7.0_75/jre

It wasn't apparent in the documentation that this should or should not work, and the second method causes 3 extra intermediate containers to be built. As the statements are very similar it would be nice to have then behave similarly, or the difference be documented.

### PR DESCRIPTION
**Pull request by duglin**
Fix some escaping in ENVs and clarify that ENV is not recursive

Closes #10391

Support for #8667 will be done in a new PR.

Signed-off-by: Doug Davis dug@us.ibm.com

### CODE REVIEW

File: builder/evaluator.go [hunk 1]
```diff
@@ -44,7 +44,7 @@ var (
 
 // Environment variable interpolation will happen on these statements only.
 var replaceEnvAllowed = map[string]struct{}{
-	"env":     {},
+	// "env":     {},    // env's func for why this is commented out
```
Comment (erikh) [inline builder/evaluator.go]: let's just delete this line, with maybe a comment at the top instead.

File: builder/dispatchers.go [hunk 2]
```diff
@@ -52,7 +60,8 @@ func env(b *Builder, args []string, attributes map[string]bool, original string)
 	for j := 0; j < len(args); j++ {
 		// name  ==> args[j]
 		// value ==> args[j+1]
-		newVar := args[j] + "=" + args[j+1] + ""
+		newValue := b.replaceEnv(args[j+1])
+		newVar := args[j] + "=" + newValue + ""
```
Comment (erikh) [inline builder/dispatchers.go]: Just checking; this is to ensure the cache string is consistent?

File: builder/evaluator.go (hunk 1 above)
Comment (duglin) [inline builder/evaluator.go]: done

File: builder/dispatchers.go (hunk 2 above)
Comment (duglin) [inline builder/dispatchers.go]: Can you elaborate? I'm not sure what you're asking.

**Comment by SvenDowideit** [comment]
is this a docs first proposal? :)

File: builder/dispatchers.go (hunk 2 above)
Comment (erikh) [inline builder/dispatchers.go]: Just confused as to what is happening, but I think I understand now. Is the string append at the end necessary?

**Comment by duglin** [comment]
I like to think of it as a bug fix. 😄

-Doug

File: builder/dispatchers.go (hunk 2 above)
Comment (duglin) [inline builder/dispatchers.go]: Been looking at that on-n-off all day trying to remember what the heck I was doing in that previous PR but I have no idea  :-)  I've removed it.

**Comment by duglin** [comment]
ping @erikh any thoughts on this?

**Comment by erikh** [comment]
I'm going to gracefully dodge this and punt to @crosbymichael instead

**Comment by crosbymichael** [comment]
Erik!! noooo

**Comment by erikh** [comment]
lol <3

**Comment by duglin** [comment]
hmmm is that supposed to be @erikh in the background? That guy jumping up and down?  :-)

**Comment by duglin** [comment]
@erikh @crosbymichael not sure how to make progress on this one.  Seems like it would be a nice addition to have this and I think its what users would expect.

**Comment by tiborvass** [comment]
@duglin do we really need this?

The user can very well just add another `ENV` instruction.

**Comment by duglin** [comment]
Once we added support for multiple var=value's per ENV statement, people expected it to act like it does on the cmd line.  Since its not a huge imposition for us to support it I think we should.  If it required us to hack stuff then I would question its value, but that's not the case.

File: integration-cli/docker_cli_build_test.go [hunk 3]
```diff
@@ -268,6 +270,11 @@ func TestBuildEnvironmentReplacementEnv(t *testing.T) {
 			if parts[1] != "foo" {
 				t.Fatalf("Could not find replaced var for env `bar`: got %q instead of `foo`", parts[1])
 			}
+		} else if strings.HasPrefix(parts[0], "env") {
+			envCount = envCount + 1
```
Comment (tiborvass) [inline integration-cli/docker_cli_build_test.go]: @duglin `envCount++` :)

Comment (duglin) [inline integration-cli/docker_cli_build_test.go]: changed

**Comment by duglin** [comment]
Any more thoughts on this one?
@jfrazelle @crosbymichael @tiborvass

**Comment by tiborvass** [comment]
Fine... design OK.

**Comment by jfrazelle** [comment]
deisgn ok, still not sure that it entirely necessary but ok :smiling_imp:

**Comment by jfrazelle** [comment]
LGTM

**Comment by thockin** [comment]
How does one specify a literal "$" in a value?

**Comment by tiborvass** [comment]
@thockin should be `\$`, and thanks I'll test it with this PR. Let me know if you have problems doing that.

**Comment by jfrazelle** [comment]
we should have a test for that

**Comment by thockin** [comment]
Similarly, how do I specify a literally quoted value?  Any characters that
have special meaning probably need to support escapes and tests should
cover those :)

**Comment by duglin** [comment]
Yes.
```
FROM ubuntu
ENV foo=\$bar
RUN env
```
shows $bar
Not sure I saw a testcase for it but I'll add one.

**Comment by duglin** [comment]
@thockin can you give me an example of what you'd like to see tested (aside from the $ one) ?

**Comment by thockin** [comment]
something like
    ENV is_quoted="no"
vs
    ENV is_quoted=\"yes\"

I don't know if there are any other special characters?  Single quotes
maybe?  Curly braces probably depend on $, so no need to explicitly handle
those.

**Comment by duglin** [comment]
@thockin see https://github.com/docker/docker/pull/10908

File: integration-cli/docker_cli_build_test.go [hunk 4]
```diff
@@ -243,6 +243,7 @@ func TestBuildEnvironmentReplacementEnv(t *testing.T) {
   FROM scratch
   ENV foo foo
   ENV bar ${foo}
+  ENV env1=$foo env2=${env1} env3="$env2" env4="${env3}"
```
Comment (tiborvass) [inline integration-cli/docker_cli_build_test.go]: @duglin Can we add a test case for `'$env'` and make sure it does not get substituted?

**Comment by tiborvass** [comment]
@duglin your example:
```
FROM ubuntu
ENV foo=\$bar
RUN env
```
is a false positive, because `bar` is not defined.

Try:
```
FROM ubuntu
ENV bar=baz foo=\$bar
RUN env
```

I put it in one line because of this PR, but the same is true for two separate ENV lines.

Interestingly, `\\$bar` properly escapes the `$` sign. Not sure if that's what was intended.

**Comment by duglin** [comment]
@tiborvass I'm not following.  In the example:
```
FROM ubuntu
ENV foo=\$bar
RUN env
```
I doesn't matter if "bar" is defined or not, we're testing to make sure that "$bar" is shown, not bar's value (which is empty).  Or are you suggesting that thing may act differently if the variable is empty vs when it has a value?

**Comment by duglin** [comment]
@tiborvass as for the \\ thing.... that might be bug, let me investigate it more.

**Comment by tiborvass** [comment]
@duglin 
> Or are you suggesting that thing may act differently if the variable is empty vs when it has a value?

Yes :)

Compare it to
```
FROM ubuntu
ENV bar=baz
ENV foo=\$bar
RUN env
```

I expect the output to show `bar=baz` and `foo=$bar`, not `foo=baz`.

**Comment by duglin** [comment]
oh my - I just checked and you’re right - it does act differently - I’m gonna have to fix this.  I’m busy with stuff right now but hopefully tomorrow afternoon I should have time to clean this up.

-Doug

**Comment by duglin** [comment]
Ok updated and rebased.
I made it
- so the ENV commands can be recursive (e.g. ENV foo=bar jjj=$foo)
- we properly handle escaping (\\) properly
- and we now support ${xxx:+...} and ${xxx:-...} syntax - see #8667 

Far more than I expected in this PR but it kind of kept going as I kept trying to "do it right" :-)

**Comment by tiborvass** [comment]
@duglin Sorry but can we not bundle the 3rd list item in this PR? Especially since I'm not sure we want that advanced substitution :(

**Comment by duglin** [comment]
@tibor - ok I removed the super cool #{xxx:...} stuff - will open a new PR to re-enable it.

File: integration-cli/docker_cli_build_test.go (hunk 4 above)
Comment (tiborvass) [inline integration-cli/docker_cli_build_test.go]: @duglin can you add `'$env'` pretty please? :)

Comment (duglin) [inline integration-cli/docker_cli_build_test.go]: holy cow - after all that work I forgot about your original ask!  :-)  Sorry about that.
Adding it around line 3868

Comment (duglin) [inline integration-cli/docker_cli_build_test.go]: I take that back - I'll add it here.

**Comment by duglin** [comment]
@tibor ok - I added the missing test

**Comment by tiborvass** [comment]
@duglin why do we still have the shellWord stuff? Is that needed for the recursive env stuff?

**Comment by duglin** [comment]
@tiborvass As I was working on the escaping stuff I realized there was an issue with how the parser was returning the words back to the builder.  The env line_parser was processing all of the \\'s and quotes so by the time the builder got the words all it had was the raw word.  Which means when we went to replace $foo with its value, we didn't know if there as a \\ before the $ or if the entire thing was in single quotes.  We lost all of this extra info.  

So I modified it so that the line_parser just returned the words 'as is', but then I needed code to deal with the quoting and escaping - and that's what I'm doing in the shellWord stuff.  It replaces the older replaceEnv code that was added recently.

File: builder/shellParser.go [hunk 5]
```diff
@@ -0,0 +1,227 @@
...
+		sw.next()
+		name := sw.processName()
+		ch = sw.peek()
+		if ch == '}' {
+			// Normal ${xx} case
+			sw.next()
+			return sw.getEnv(name), nil
+		}
+		return "", fmt.Errorf("Unsupported ${} substitution: %s", sw.word)
+		/* comment out support for advances ${..} syntax stuff for now
```
Comment (tiborvass) [inline builder/shellParser.go]: can you remove this dead code please? You can keep it in another branch

**Comment by tiborvass** [comment]
@duglin I'd prefer if we could separate the two fixes in two different PRs. Sorry to be annoying again, but I'm like so close to merging the recursive ENV stuff, but I cannot because of new code. And the shellword stuff is not particularly straightforward to review. I'm also wondering whether we should not simply use https://github.com/mattn/go-shellwords for things that are not JSON syntax.

**Comment by duglin** [comment]
@tiborvass I removed the commented-out code.
As for using that other package... I'm not sure that's wise.  We'd have to hack it to remove support for back-ticks and in my next PR I'm planning on adding support for advanced ${} stuff - which it doesn't support.

**Comment by tiborvass** [comment]
@duglin okay, but you still don't want to put that shell stuff in a separate PR?

**Comment by duglin** [comment]
@tiborvass Yes I'm moving the ${xxx:...} stuff to a new PR.

**Comment by duglin** [comment]
@tiborvass if you mean "do I want to move the shellParser.go" to  a new PR, no I can't - I need it to do escaping properly for this PR.

**Comment by duglin** [comment]
@tiborvass any other comments?

**Comment by tiborvass** [comment]
If people could help me review this one, that'd be really great :S

I reviewed the recursive env stuff, and it looks good to me, but I'd love another pair of eyes on the shellword parsing.

Ping @tianon @LK4D4 @estesp

**Comment by estesp** [comment]
I've been looking at it some--I guess one question is what's the plan for docs on this?  Given it works "sort of like bash" but doesn't have any of the special features of bash (and yes, I know about the other PR), where will it be described the boundaries of this PR's parsing support?

**Comment by duglin** [comment]
@estesp right now I don't think we need to add any new docs since this shouldn't change existing behavior.  Once we add that other PR I'll include more docs to describe what kind of special ${} syntax we support.
Although, if you think the current docs are too slim wrt even what we support today then I can beef them up.

File: builder/shellParser.go [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+		if ch == '\000' || ch == '\'' {
+			break
+		}
+		result += string(ch)
+	}
+	return result, nil
+}
+
+func (sw *shellWord) processDoubleQuote() (string, error) {
+	// All chars up to the next " are taken as-is, even ', except $'s.
```
Comment (estesp) [inline builder/shellParser.go]: This seems like a silly thing, but I stared at this comment for awhile trying to understand why `$'` was a special sequence, until I realized (I think?) that you are trying to say "dollar signs", but using *apostrophe s* as the (incorrect) English plural of $ :) :)  Maybe for weird people like me you can change to "except any `$` characters."

File: builder/shellParser.go [hunk 7]
```diff
@@ -0,0 +1,200 @@
...
+// Process the word, starting at 'pos', and stop when we get to the
+// end of the word or the 'stopCh' character
+func (sw *shellWord) processStopOn(stopCh rune) (string, error) {
+	var charFuncMapping = map[rune]func() (string, error){
+		'\'': sw.processSingleQuote,
+		'"':  sw.processDoubleQuote,
+		'$':  sw.processDollar,
+	}
+
+	result := ""
```
Comment (estesp) [inline builder/shellParser.go]: There are a lot of initializations in this code that I don't see much in other Go code.. "var result string" seems more idiomatic to Go (http://golang.org/ref/spec#The_zero_value).

File: builder/shellParser.go (hunk 6 above)
Comment (duglin) [inline builder/shellParser.go]: LOL just for you.

**Comment by estesp** [comment]
I added a few comments to the code; one general comment--the `builder/words` tests seem good for positive tests, but I don't see much testing for error cases (where bad env strings that should error out of your parser) are tested.. e.g. things like `'` that never terminates, etc.  Seems that given parsing can be error-prone and this is much more involved than the prior, some tests around edge cases (strings with no prior quote (double or single) ending with a quote, for example) would be useful in case someone less accomplished than @duglin makes a future change here :)

File: builder/shellParser.go (hunk 7 above)
Comment (duglin) [inline builder/shellParser.go]: Old habits die hard  :-)
Changed

**Comment by estesp** [comment]
Oh, and to respond to my initial comment and your response--I was confused; I had no idea `${}` was already supported in `ENV` as I've only seem simple cases in Dockerfiles mostly.. so you are correct that at this point, the docs are probably not in need of any significant update.  It does seem that some note at the next release about ENV substitution in the same line now works as it could be assumed that nothing has changed unless someone who saw it fail up through 1.5.0 tries again in a future release.

**Comment by duglin** [comment]
@estesp there's only one reason right now that the parser returns an error, and that's when there's a syntax error in ${} - e.g. a missing }.  I added a test for that.  As for the other situations, I tried to be as forgiving as possible; so on things like a missing quote I let those go through.  I prefer to be more lenient on what we accept, but if people feel strongly that we should be as strict as possible I don't mind throwing an error.

Pushed changes for other comments.

**Comment by tianon** [comment]
I don't think I agree with this one:
```console
$ export a='LOL' b="$a"
$ echo $b

$ 
```

**Comment by duglin** [comment]
hmm @n0coast any thoughts on this given you opened #10391?
To be honest I'm torn on this one.  My initial preference is to do what the shell does, which it appears means that values aren't available until _after_ the current command is run. Which means #10391 would be an invalid usecase.
On the other hand, given ENV commands create layers, allowing for this optimization could be good.
If I had to choose right now, I'd go with what the shell does as @tianon suggests.  
But I'd like to hear other people's opinion on this.

**Comment by duglin** [comment]
Well, this is kind of interesting:
```
#!/bin/bash
ABC=ABC DEF=$ABC
echo ABC=$ABC
echo DEF=$DEF
```
yields:
```
ABC=ABC
DEF=ABC
```

**Comment by duglin** [comment]
@tianon same result from cmd line too, bash and sh
Guess it depends on whether ENV is the same as 'export' or just an assignment.
Given this, I'm now leaning the other way

**Comment by n0coast** [comment]
It does make sense that values aren't available until after the command is run, but it would be cool to do it inline to reduce layers... so that doesn't add much to the conversation since you said that already. What I wanted out of the request was more specific documentation, since when I tried doing it inline nothing said I could or could not do it, but you are allowed to set multiple values in one ENV statement... Probably the easier way would be to update documentation to say variables cannot inherit values if set using inline ENV commands.

**Comment by tiborvass** [comment]
> Given this, I'm now leaning the other way

@duglin so which way?

**Comment by duglin** [comment]
Since:
ABC=foo DEF=$ABC
is valid for sh/bash scripts I think we should support it.

**Comment by estesp** [comment]
Given the responses to my original Qs, and the confirmation on matching the "bash-isms" (ignoring that `export` seems to have its own behavior), I'm **LGTM** on this one.

**Comment by duglin** [comment]
@tianon you ok with this?

**Comment by tianon** [comment]
As we discussed on IRC (https://botbot.me/freenode/docker-maintainers/msg/34045583/), I'm still -1 on this because `ENV` as currently implemented really _is_ `export`.

**Comment by duglin** [comment]
@tianon I updated it last night to act like 'export' - so I removed the recursiveness

**Comment by tianon** [comment]
@duglin :+1:!

SGTM

File: builder/shellParser.go [hunk 8]
```diff
@@ -0,0 +1,201 @@
...
+	return sw.process()
+}
+
+func (sw *shellWord) process() (string, error) {
+	return sw.processStopOn('\000')
+}
+
+// Process the word, starting at 'pos', and stop when we get to the
+// end of the word or the 'stopCh' character
+func (sw *shellWord) processStopOn(stopCh rune) (string, error) {
```
Comment (tiborvass) [inline builder/shellParser.go]: @duglin can you explain this `stopCh` and `\000` ?

Comment (tiborvass) [inline builder/shellParser.go]: oh you're using 0 as the end of string

Comment (tiborvass) [inline builder/shellParser.go]: @duglin why not remove `process()` and rename `processStopOn(rune)` to `process()` ?

Comment (duglin) [inline builder/shellParser.go]: Well, before I split this PR into two processStopOn was recursive to handle ${}'s in ${}.  So, while it may not be needed right now, I'd prefer to keep it like it is because #11149 will use it.

Comment (tiborvass) [inline builder/shellParser.go]: @duglin okay then i'd prefer `stopChar`, `stopCh` feels like a channel

Comment (duglin) [inline builder/shellParser.go]: done!

**Comment by jfrazelle** [comment]
ping @tiborvass @duglin have you all worked this out?

**Comment by duglin** [comment]
@jfrazelle I think its done - just waiting for @tiborvass to give his ok.
rebased just for fun and to force a rebuild

File: builder/parser/line_parsers.go [hunk 9]
```diff
@@ -107,12 +107,15 @@ func parseNameVal(rest string, key string) (*Node, map[string]bool, error) {
 				quote = ch
 				blankOK = true
 				phase = inQuote
-				continue
```
Comment (tiborvass) [inline builder/parser/line_parsers.go]: @duglin it's not clear to me anymore if it's this function's or the shellParser's responsibility to handle quotes.

Comment (duglin) [inline builder/parser/line_parsers.go]: @tiborvass if you look at https://github.com/duglin/docker/blob/Issue10391/builder/parser/line_parsers.go#L109 you'll see that when we see a quote we change our "phase" and we save the quote character (in the 'quote' variable), but then we fall thru to https://github.com/duglin/docker/blob/Issue10391/builder/parser/line_parsers.go#L109 which adds the current character ( which is the quote) to the word.
So, short answer, shellParser will remove the quotes. We needed to move it from here because later on when we do the ${} processing we have to know whether there are single vs double quotes around things to know whether to do the substitution. So this func is now just responsible for breaking things into words.

File: builder/dispatchers.go [hunk 10]
```diff
@@ -53,7 +53,6 @@ func env(b *Builder, args []string, attributes map[string]bool, original string)
 		// name  ==> args[j]
 		// value ==> args[j+1]
 		newVar := args[j] + "=" + args[j+1] + ""
-		commitStr += " " + newVar
```
Comment (tiborvass) [inline builder/dispatchers.go]: can you explain this?

Comment (duglin) [inline builder/dispatchers.go]: Good catch and I have no idea why it was removed - I don't recall being in that code since its about LABEL and not ENV.  Very odd.  It does make me wonder why the tests didn't fail though when I was removed.  I'll work on adding a test to catch that.

Comment (duglin) [inline builder/dispatchers.go]: just checked - and there are no tests of the cache w.r.t. LABELs.  scary!  :-)   will work on that.

**Comment by jfrazelle** [comment]
ping @tiborvass for the merge if he approves :)

File: docs/sources/reference/builder.md [hunk 11]
```diff
@@ -146,6 +146,17 @@ The instructions that handle environment variables in the `Dockerfile` are:
 `ONBUILD` instructions are **NOT** supported for environment replacement, even
 the instructions above.
 
+Environment variable subtitution will use the same value for each variable
+throughout the entire command.  In other words, in this example:
+
+	ENV abc=hello
+    ENV abc=bye def=$abc
```
Comment (tiborvass) [inline docs/sources/reference/builder.md]: indentation

File: integration-cli/docker_cli_build_test.go [hunk 12]
```diff
@@ -2128,7 +2147,7 @@ func TestBuildRelativeWorkdir(t *testing.T) {
 
 func TestBuildWorkdirWithEnvVariables(t *testing.T) {
 	name := "testbuildworkdirwithenvvariables"
-	expected := "/test1/test2/$MISSING_VAR"
```
Comment (tiborvass) [inline integration-cli/docker_cli_build_test.go]: why?

File: integration-cli/docker_cli_build_test.go [hunk 13]
```diff
@@ -239,9 +239,18 @@ func TestBuildEnvironmentReplacementEnv(t *testing.T) {
 
 	_, err := buildImage(name,
 		`
-  FROM scratch
+  FROM busybox
   ENV foo foo
   ENV bar ${foo}
+  ENV abc1='$foo'
+  ENV env1=$foo env2=${foo} env3="$foo" env4="${foo}"
+  RUN [ "$abc1" = '$foo' ] && (echo "$abc1" | grep foo)
```
Comment (tiborvass) [inline integration-cli/docker_cli_build_test.go]: why grep?

File: builder/parser/line_parsers.go [hunk 14]
```diff
@@ -122,15 +125,20 @@ func parseNameVal(rest string, key string) (*Node, map[string]bool, error) {
...
 			}
-			if ch == '\\' {
+			// \ is special except for ' quotes - can't escape anything for '
+			if ch == '\\' && quote != '\'' {
 				if pos+1 == len(rest) {
 					phase = inWord
 					continue // just skip \ at end
 				}
+				// If next char is " then don't add the \ - its just escaping "
```
Comment (tiborvass) [inline builder/parser/line_parsers.go]: @duglin can you please remove this along with the dead code? Even if it's useful for another PR. Also, why would we remove the backslash anyway? `echo 'hello\"world'` yields `hello\"world`.

File: builder/parser/testfiles/env/result [hunk 15]
```diff
@@ -2,9 +2,14 @@
...
+(env "name" "value\\ value2")
+(env "name" "\"value'quote space'value2\"")
+(env "name" "'value\"double quote\"value2'")
+(env "name" "value\\ value2" "name2" "value2\\ value3")
+(env "name" "\"a\\\"b\"")
+(env "name" "\"a\\'b\"")
+(env "name" "'a\\'b'")
+(env "name" "'a\\'b''")
+(env "name" "'a\\\"b'")
+(env "name" "value" "name1" "value1" "name2" "\"value2a            value2b\"" "name3" "\"value3a\\n\\\"value3b\\\"\"" "name4" "\"value4a\\\\nvalue4b\"")
```
Comment (erikh) [inline builder/parser/testfiles/env/result]: might be useful to test `"''"` here. (that's a double quote, two single quotes, and another double)

File: builder/shellParser_test.go [hunk 16]
```diff
@@ -0,0 +1,51 @@
+package builder
```
Comment (erikh) [inline builder/shellParser_test.go]: why is this in the builder and not in the parser?

File: builder/words [hunk 17]
```diff
@@ -0,0 +1,42 @@
+hello                    |     hello
```
Comment (erikh) [inline builder/words]: same here

File: builder/shellParser.go [hunk 18]
```diff
@@ -0,0 +1,201 @@
...
+			if err != nil {
+				return "", err
+			}
+			result += tmp
+		} else {
+			ch = sw.next()
+			if ch == '\\' {
+				chNext := sw.peek()
+
+				if chNext == '\000' {
```
Comment (tiborvass) [inline builder/shellParser.go]: @duglin this feels like duplicate work again. I think we should remove the end-of-line `\` in the first parser. It should remove both the `\` and the `\n`, that way it's truly as-if there were no newlines. shellWord should just get the whole thing, and be assured that there is no `\` at the end of a line and you could thus remove this logic.

Question would be then: what happens when there's a `\` at the very end of a Dockerfile (without `\n`) ?

File: docs/sources/reference/builder.md (hunk 11 above)
Comment (duglin) [inline docs/sources/reference/builder.md]: fixed

File: integration-cli/docker_cli_build_test.go (hunk 13 above)
Comment (duglin) [inline integration-cli/docker_cli_build_test.go]: I needed to make sure that $foo wasn't replaced so grep'ing for 'foo' seemed like a good test.

File: builder/parser/line_parsers.go (hunk 14 above)
Comment (duglin) [inline builder/parser/line_parsers.go]: @tiborvass why would I remove this? Its not dead code.  Note that line 135 is hit when we're in double-quotes not single.

File: builder/parser/line_parsers.go [hunk 19]
```diff
@@ -122,15 +125,20 @@ func parseNameVal(rest string, key string) (*Node, map[string]bool, error) {
...
 				pos++
-				ch = rune(rest[pos])
+				nextCh := rune(rest[pos])
+				// if nextCh != '"' {
```
Comment (tiborvass) [inline builder/parser/line_parsers.go]: @duglin this is the dead code i was referring to

File: builder/parser/testfiles/env/result (hunk 15 above)
Comment (duglin) [inline builder/parser/testfiles/env/result]: Done

File: integration-cli/docker_cli_build_test.go (hunk 13 above)
Comment (tiborvass) [inline integration-cli/docker_cli_build_test.go]: @duglin if $foo were replaced it would be replaced to `foo`.

File: builder/parser/line_parsers.go (hunk 19 above)
Comment (duglin) [inline builder/parser/line_parsers.go]: Ah, yes that's dead - just forgot to remove it after testing

File: builder/words (hunk 17 above)
Comment (duglin) [inline builder/words]: @erikh I assume you mean to add a "''" test, right?  if so, done.

File: builder/shellParser_test.go (hunk 16 above)
Comment (duglin) [inline builder/shellParser_test.go]: not 100% sure what I was thinking at that exact moment :-)  but I probably put it in the builder because shellParser (which is probably not the best name for it) isn't used during the parsing phase - its actually used during the env-var substitution phase.  Perhaps I should have called it "shellWordProcessor" instead - dunno.

File: builder/words (hunk 17 above)
Comment (erikh) [inline builder/words]: No, i'm wondering if there's a reason this code exists in the builder instead of the parser. Is this a runtime evaluation thing?

Sorry if I'm behind; it's been a very crazy month.

-Erik

File: builder/shellParser_test.go (hunk 16 above)
Comment (erikh) [inline builder/shellParser_test.go]: Nah, it's fine, I was just confused. Thanks for responding. No objections from me.

-Erik

File: integration-cli/docker_cli_build_test.go (hunk 13 above)
Comment (duglin) [inline integration-cli/docker_cli_build_test.go]: doi!  replacing "foo" with "zzz"

File: builder/shellParser_test.go (hunk 16 above)
Comment (erikh) [inline builder/shellParser_test.go]: Actually I have one minor objection. I don't think the project uses camelCase anywhere for filenames and I wouldn't want to rock the boat. :P

-Erik

File: builder/words (hunk 17 above)
Comment (duglin) [inline builder/words]: No, its a build-time thing but its not a parser thing, its more like a cmd interpreter thing - so its after the parser.

File: builder/shellParser_test.go (hunk 16 above)
Comment (duglin) [inline builder/shellParser_test.go]: old habits die hard.  I'll rename it to shell_parser...

File: builder/shellParser.go (hunk 18 above)
Comment (duglin) [inline builder/shellParser.go]: Which is why I don't want to remove the logic.  I wouldn't want this code to assume too much about what it is given.  I would much prefer to handle generic words gracefully.  If its never given that case then no harm done, but I'd prefer for this code to be stand-alone bullet-proof, as much as possible anyway.  Part of me wouldn't mind moving it into a pkg.

File: builder/shellParser.go [hunk 20]
```diff
@@ -0,0 +1,201 @@
...
+		name += string(ch)
+	}
+
+	return name
+}
+
+func (sw *shellWord) getEnv(name string) string {
+	for _, env := range sw.envs {
+		i := strings.Index(env, "=")
+		if i < 0 || name != env[:i] {
```
Comment (tiborvass) [inline builder/shellParser.go]: @duglin what happens if there's no `=` ? that is `i < 0`... it should either panic, or mean empty string, in which case it should `return ""`.

Comment (duglin) [inline builder/shellParser.go]: I don't think we'll ever have that case due to how the env array is setup in the builder, but I like the idea of being extra careful so I'll make it return "" in that case, I think its safe to assume that
var=
and
var
should be treated the same - just replace with empty string.

File: builder/parser/line_parsers.go (hunk 14 above)
Comment (tiborvass) [inline builder/parser/line_parsers.go]: @duglin this comment is now irrelevant

Comment (duglin) [inline builder/parser/line_parsers.go]: oops yup -removed it

File: integration-cli/docker_cli_build_test.go [hunk 21]
```diff
@@ -239,9 +239,18 @@ func TestBuildEnvironmentReplacementEnv(t *testing.T) {
...
 	_, err := buildImage(name,
 		`
-  FROM scratch
-  ENV foo foo
+  FROM busybox
+  ENV foo zzz
   ENV bar ${foo}
+  ENV abc1='$foo'
+  ENV env1=$foo env2=${foo} env3="$foo" env4="${foo}"
+  RUN [ "$abc1" = '$foo' ] && (echo "$abc1" | grep foo)
```
Comment (tiborvass) [inline integration-cli/docker_cli_build_test.go]: `grep -q`

Comment (duglin) [inline integration-cli/docker_cli_build_test.go]: done

File: integration-cli/docker_cli_build_test.go (hunk 12 above)
Comment (tiborvass) [inline integration-cli/docker_cli_build_test.go]: For reference: this means that we are making a breaking change: undefined variables are now substituted as expected. The rationale for this old behavior was to maintain backwards compatibility when we introduced ENV var substitution. Now that ENV var substitution is expected, it is much more confusing to keep it, since people expect it to be replaced by an empty string.

**Comment by duglin** [comment]
@tiborvass can we please resolve this one today so it can make 1.6 ?

**Comment by tiborvass** [comment]
Yes. also this is a bug fix so it doesn't have to make it today but I'll
merge it anyways.

**Comment by jfrazelle** [comment]
If you are going to merge it can it be done in the next hour or so, so it
will be included in the RC for testing

**Comment by tiborvass** [comment]
LGTM

**Comment by tiborvass** [comment]
@moxiegirl @SvenDowideit @fredlf If you need to make adjustments, please do so in a new PR, thanks! :)

**Comment by duglin** [comment]
@tiborvass thanks!
